### ISSUE TITLE
A different approach leveraging async iterators 

### ISSUE THREAD

**Issue by Jonasdoubleyou** [isiahmeadows/proposal-promise-schedule#2]
As already presented on [ES Discourse](https://es.discourse.group/t/proposal-seeking-champion-composable-promise-concurrency-management/713/8), I think that the same problem space ("scheduling tasks concurrently")  can also be tackled from another angle, leveraging (async) iterators to compose streams of asynchronous tasks. 

_This issue is meant as a source of inspiration and food for thought. It is by no means as thought out as the proposal is, and is purely hypothetical._

JavaScript currently has *synchronous iterators* and *asynchronous iterators*. In general an iterator can be _consumed_ by calling it's `.next()` method (repeatedly, possibly with arguments), and it _produces_ values by returning values to the caller, e.g. by `yield`ing inside a generator function. The term _producer_ will be used in the following to describe an iterator who produces values, but does not expect any arguments in it's `.next` method (in the same way a _consumer_ could be defined). 

**AsyncIterator.fromProducer**

For a synchronous iterator, consumption and production are directly coupled, for an asynchronous iterator however, where each iteration returns a Promise, the Promise will be settled at a later point in time, thus consumption will happen before production, 
and eventually an asynchronous iterator can be consumed multiple times before production happens (for asynchronous generators the specification already defines a queue to deal with that situation). This decouples the _production event_ from the consumption, so it would also be possible to define a function, which creates a callback and an asynchronous iterator, such that whenever the callback gets called, the asynchronous iterator advances:

```js
const clicks = AsyncIterator.fromProducer(produce => {
  window.onclick = produce;
}); 

clicks
  .forEach(console.log); 
```

This works for every source of events because events are always ordered by time, and as such can be represented as an ordered iteration. There are however two open questions: 

- What happens if a production event happens before the iterator gets consumed? (I'd say it throws, and leaves buffering to `.parallel`)
- When does the iterator close? (a separate callback, calling `produce(value, /* close: */ true)` , though other variants are possible)

A brief implementation would be:

```js
AsyncIterator.fromProducer = async function* fromProducer(producer) {
  const abort = function() { throw new Error("Producer is not consuming"); };
  let resolve = abort;
  producer(v => resolve(v)); 
  while(true) {
    const result = await new Promise(r => resolve = r);
    resolve = abort;
    yield result;
  }
};
```

**Iterator<Promise<T>>.parallel(max: number = Infinity): AsyncIterator<T>**

To consume an (async) iterator, perform an asynchronous task for each element, and then wait for it to finish can already be done with `Promise.all`:

```js
await Promise.all(
  iterator
   .map(asyncTask)
   .toArray()
)
```

Now while that already works this way for a lot of cases, with that we still have the two problems described in this proposal:

> Do you have a potentially very large or unbounded sequence and you want to process it concurrently? Congratulations, Promise.all will never work in your favor and you will have to use the long form with an explicit open counter to not run out of memory.

> Do you need to limit concurrency to ensure throughput or limit memory usage? You're forced to either do the longest form or manually iterate the collection with occasional await readyForNextValue() calls in the loop, as Promise.all doesn't offer anything to let you manage concurrency.

To represent concurrent tasks, we can use some sort of collection of Promises, so if we have an iterator that returns Promises, we can start a task by consuming the iterator, and it ends when the Promise settles. Unlike an async iterator where the iterator eventually stops, it is synchronously known whether there is a next task, however the result of the task eventually arrives. A method could be created onto an iterator (the _origin_), which returns an async iterator. When that iterator is consumed, the origin gets consumed till the end. For each Promise from the origin, when it resolves it's value gets stored in a buffer, or when the iterator is currently consumed it resolves with that value. When the iterator gets consumed again, _some_ value from the buffer is taken and the iterator produces a result at the next tick, or when the buffer is empty the Promise stays pending. If "overconsumption" happens, the same procedure happens as with async generators (though consumers should usually not do this). 

With that, the async iterator `.parallel()` returns the following attributes:

- It holds a buffer which might grow as large as the origin produces values
- It "get's as much as it can get" from the origin
- When overconsumed, the order of values is ascending by resolution time, otherwise the order is not defined (but might be defined later)
- It can be consumed at a different speed than the origin produces values and will either way either buffer values or Promises
- Just calling `.parallel` does not consume the origin, only the first (and further) `.next` calls do

With that the above example could be rewritten as (though unlike `Promise.all` order is not maintained, which in a lot of cases is irrelevant anyways):

```js
await iterator
  .map(asyncTask)
  .parallel()
  .toArray());
```

Now that still does not solve the problems described above, though it does when the `max` parameter get's introduced: 
 When consuming the origin for the first time, only consume `max` iterations, then, whenever a Promise resolves and the iterator is not fully consumed, consume another entry. With that, limiting e.g. the parallel calls to an external API turns as easy as:

```js
await locations.values()
  .map(fetchWeatherAPI)
  .parallel(10)
  .toArray()
```

The described buffer could also be sorted in a future extension, allowing for prioritization of tasks. 

**AsyncIterator<Promise<T>>.parallel(max: number = Infinity): AsyncIterator<T>**

Expanding the `.parallel()` function to async iterators should be straightforward, though it introduces even more concurrency (which is great cause that's exactly what this proposal tackles). The specification currently recommends not to put a Promise into an asyc iterator's value, though I'd say that the information "a next value exists" and "the next value" can arrive at different points in time, especially when mapping an async iterator with an async function. 

There is one caveat though, the current [`.map` proposal](https://tc39.es/proposal-iterator-helpers/#sec-asynciteratorprototype.map) does flatten promises, so `asyncIterator.map(asyncFn)` does not produce Promises. 

```js
AsyncIterator.fromProducer(addLocation => { /* ... */ })
  .parallel() // if capacity of "fetchWeatherAPI" is reached, buffer here
  .map(fetchWeatherAPI)
  .parallel() // -> This doesn't work!  This will create a lot of buffered Promises, but fetchWeatherAPI will be called in sequence, not in parallel
  .next();
```

I'm not quite sure how to resolve this conflict, but as an alternative variant, `parallel` could take a function as a parameter instead which will be called for each element, similar as `.map`. To not complicate this issue with even more variants, the following uses an imaginary method `.asyncMap` which does not have step f described in the proposal. 

**Concrete examples**

1.) Limited concurrency on a list of tasks:

```js
// Current proposal
const results = await runAndCollect(schedule => {
    for (const item of list) {
        schedule(() => callApi({id: item.id, ...}))
    }
}, {
    maxConcurrency: 10,
    onResolved() { advanceProgressBar() },
})

// This idea
list.values()
  .map(callApi)
  .parallel(10)
  .forEach(advanceProgressBar);
```

2.) Consuming all asynchronous task results as an array:

```js
// Current proposal
async function runAndCollect(init, opts) {
    const results = []

    await Promise.scheduleAndRun(init, {
        ...opts,
        onResolved(result) {
            results.push(result)
            if (opts && opts.onResolved) opts.onResolved(result)
        },
        onRejected(error) {
            if (opts && opts.onRejected) opts.onRejected(error)
            throw error
        },
    })

    return results
}

// This idea
const results = await AsyncIterator.fromProducer(init)
  .asyncMap(it => it())
  .parallel(10)
  .toArray()
```

3.) Reading files, processing them, then writing to another location with buffering inbetween:

```js
filenames.values()
  .map(readFile)
  .parallel()
  .asyncMap(process)
  .parallel()
  .asyncMap(writeFile)
  .forEach(console.log);
```

**Summary**

Compared to the existing proposal I'd say that this variant is more composable and as such covers more usecases (hypothetically). It extends the asynchronous iterators and makes them more useful (cause they are still a very rarely used feature), however this also makes the whole thing far more complex and decoupling the different parts might make it more difficult to optimize for engines. Given that the problems described are caused by heavy workload, weighting performance over composability might be a valid argument for the existing proposal.  The methods presented here are also "purely academic" and are not based on real usage patterns, I also didn't actually test them yet ...

**Comment by dead-claudia** [comment]
There's a catch: async iterators are still sequential, and [the async iterator proposal plans to keep it that way](https://tc39.es/proposal-iterator-helpers/#sec-asynciteratorprototype.map).

```js
"use strict"

async function *emit() {
    console.log("emit a")
    await Promise.resolve()
    console.log("emit resolve a")
    yield "a"
    console.log("emit b")
    await Promise.resolve()
    console.log("emit resolve b")
    yield "b"
    console.log("emit c")
    await Promise.resolve()
    console.log("emit resolve c")
    yield "c"
    console.log("emit done")
}

const gen = emit()
console.log("next 1")
gen.next().then(v => console.log(`1: ${v.value}`))
console.log("next 2")
gen.next().then(v => console.log(`2: ${v.value}`))
console.log("next 3")
gen.next().then(v => console.log(`3: ${v.value}`))
console.log("next 4")
gen.next().then(v => console.log(`4: ${v.value}`))
console.log("done")

// Logs:
// next 1
// emit a
// next 2
// next 3
// next 4
// done
// emit resolve a
// emit b
// 1: a
// emit resolve b
// emit c
// 2: b
// emit resolve c
// emit done
// 3: c
// 4: undefined
```

This means you can't make it parallel unless you change those async iterators work to support concurrent calls like that.

Not that I'm against it, but it's something that [should definitely be addressed in that proposal](https://github.com/tc39/proposal-iterator-helpers).

**Comment by dead-claudia** [comment]
Also, uncontrolled parallelism isn't possible to manage - you don't have any info on whether the iterator completes or not until the call returns.

**Comment by Jonasdoubleyou** [comment]
Yeah, I think it works quite nicely for synchronous iterators returning Promises, but with asynchronous iterators this runs into various conflicts with the existing design, especially when chaining multiple `.asyncMap`s ... An alternative approach which would avoid that conflict would be combining the proposed semantics for `.parallel` and map into a `.parallelMap`:

```
filenames.values()
  .parallelMap(readFile)
  .parallelMap(process)
  .parallelMap(writeFile)
  .forEach(console.log);
```

That should work equally well for both synchronous and asynchronous iterators, as the promising task get's moved out of the iterator (instead of `AsyncIterator<Promise<T>>.parallel()` it'll be `AsyncIterator<S>.parallelMap(map: (value: S) => Promise<T>): AsyncIterator<T>`. 

Actually looking at the examples above, I always used the `.map(..).parallel()` pattern, so it would be feasible to combine both into one. Only in one case I used `AsyncIterator.fromProducer(...).parallel()` to enable buffering, though actually that's quite a weird usage of `.parallel()` as actually no Promises are being produced (basically this is more just "buffering" than "buffering of concurrently running task results"). 

When I just tried to implement `.parallelMap` based on `.Promise.scheduleAndRun` (to compare both) I noticed one quite large limitation with `scheduleAndRun`: As the name already suggests, one can only _actively schedule_ new tasks, whereas `.parallelMap` can "ask for new tasks" by further consuming the (async) origin. Especially when processing tasks based on asynchronous resources (files, external systems, ...) this might be a quite common usecase.

**Comment by dead-claudia** [comment]
That's true, but a `parallelMap` still doesn't let you schedule new tasks based on existing tasks. [Here's an example in the wild where a parallel map wouldn't have worked.](https://github.com/awslabs/aws-api-gateway-developer-portal/blob/18e152bdbb398358b15e0b713862fb6d9a3cba95/lambdas/catalog-updater/index.js#L189-L267) If I translate that to use my proposed `Promise.scheduleAndRun`, that complicated bit becomes considerably easier to understand (and also 42 instead of 70 lines of code excluding whitespace) while still retaining the performance boost of doing the folder iteration concurrently with the usage plans loading.

```js
const handler = async (event) => {
  console.log(`event: ${inspect(event)}`)

  let s3Request

  await Promise.scheduleAndRun(async schedule => {
    const builderPromise = schedule(async () => {
      const usagePlansPromise = getAllUsagePlans(exports.apiGateway)
      const response = await exports.s3.getObject({ Bucket: process.env.BucketName, Key: 'sdkGeneration.json' }).promise()
      const sdkGeneration = JSON.parse(response.Body.toString())
      console.log(`sdkGeneration: ${inspect(sdkGeneration)}`)
      const usagePlans = await usagePlansPromise
      console.log(`usagePlans: ${inspect(usagePlans)}`)
      return new CatalogBuilder(usagePlans, sdkGeneration)
    })

    let token
    do {
      s3Request = exports.s3.listObjectsV2(
        token != null
          ? { Bucket: process.env.BucketName, Prefix: 'catalog/', ContinuationToken: token }
          : { Bucket: process.env.BucketName, Prefix: 'catalog/' }
      )
      const listObjectsResult = await s3Request.promise()
      for (const file of listObjectsResult.Contents) {
        if (exports.swaggerFileFilter(file)) {
          schedule(async () => {
            const s = await getSwaggerFile(file)
            ;(await builderPromise).addToCatalog(s)
          })
        }
      }
      if (!listObjectsResult.IsTruncated) break
      token = listObjectsResult.NextContinuationToken
    } while (s3Request != null)
  }, {
    onRejected(e) {
      if (s3Request != null) {
        s3Request.abort()
        s3Request = null
      }
      throw e
    }
  })
}
```

> Note: I revised the proposal to return a promise to the task result from `schedule` instead of a simple boolean noting if the concurrency limit was hit.

**Comment by dead-claudia** [comment]
More to the point, it would've worked in theory, but it would've taken a lot more code to stub out.

**Comment by Jonasdoubleyou** [comment]
If I'm not mistaken, `Promise.scheduleAndRun` could be implemented using a combination of `AsyncIterator.fromProducer` and `.parallelMap`:

```js
Promise.scheduleAndRun = async function(producer, options) {
  await AsyncIterator.fromProducer(producer)
    .parallelMap(it => it().then(result => ({ result }), error => ({ error })), options.maxConcurrency)
    .forEach(({ result, error }) => {
       if(result) options.onResolved(result);
       if(error) options.onRejected(error);
     });
}
```

**Comment by dead-claudia** [comment]
@Jonasdoubleyou You could use that, but it's a bit more complicated to come up with and considerably less straightforward (I had to recast multiple parts of this to make it fit).

```js
const handler = async (event) => {
  console.log(`event: ${inspect(event)}`)

  let s3Request

  const builderPromise = (async () => {
    const usagePlansPromise = getAllUsagePlans(exports.apiGateway)
    const response = await exports.s3.getObject({ Bucket: process.env.BucketName, Key: 'sdkGeneration.json' }).promise()
    const sdkGeneration = JSON.parse(response.Body.toString())
    console.log(`sdkGeneration: ${inspect(sdkGeneration)}`)
    const usagePlans = await usagePlansPromise
    console.log(`usagePlans: ${inspect(usagePlans)}`)
    return new CatalogBuilder(usagePlans, sdkGeneration)
  })()

  try {
    await AsyncIterator.from((async function *() {
      let token
      do {
        s3Request = exports.s3.listObjectsV2(
          token != null
            ? { Bucket: process.env.BucketName, Prefix: 'catalog/', ContinuationToken: token }
            : { Bucket: process.env.BucketName, Prefix: 'catalog/' }
        )
        const listObjectsResult = await s3Request.promise()
        yield* listObjectsResult.Contents
        if (!listObjectsResult.IsTruncated) break
        token = listObjectsResult.NextContinuationToken
      } while (s3Request != null)
    })())
      .parallelMap(async file => {
        if (exports.swaggerFileFilter(file)) {
          const s = await getSwaggerFile(file)
          ;(await builderPromise).addToCatalog(s)
        }
      })
      .forEach(() => {})
  } catch (e) {
    if (s3Request != null) {
      s3Request.abort()
      s3Request = null
    }
    throw e
  }
}
```
